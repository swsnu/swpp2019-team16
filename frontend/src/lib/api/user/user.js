import client from '../client';

export function get({ id }) {
  return client.get(`/api/v1/user/${id}`);
}

export function updatePoint({ userId, point }) {
  return client.put(`/api/v1/user/${userId}`, { point });
}

export function riderOnTaxi({ riderId }) {
  return client.put(`/api/v1/user/taxi/rider/${riderId}`);
}

export function driverGoTaxi({ driverId }) {
  return client.put(`/api/v1/user/taxi/driver/${driverId}`);
}
