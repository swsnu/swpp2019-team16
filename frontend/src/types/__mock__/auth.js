export const mockRiderAuth = {
  id: 1,
  email: 'rider1@gmail.com',
  user_type: 'rider',
  vehicle: {
    car_type: '',
    plate_no: '',
  },
};

export const mockDriverAuth = {
  id: 1,
  email: 'driver1@gmail.com',
  user_type: 'driver',
  vehicle: {
    car_type: '',
    plate_no: '',
  },
};
