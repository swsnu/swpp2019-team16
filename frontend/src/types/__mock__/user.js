export const mockRider = {
  id: 1, // rider/driver id
  status: 'IDLE',
  group: 1, // group id
  user: {
    id: 2,
    email: 'rider1@gmail.com',
    user_type: 'rider',
    vehicle: {
      car_type: '',
      plate_no: '',
    },
  },
};

export const mockDriver = {
  id: 1, // rider/driver id
  status: 'IDLE',
  group: 1, // group id
  user: {
    id: 2,
    email: 'driver1@gmail.com',
    user_type: 'driver',
    vehicle: {
      car_type: 'benz',
      plate_no: '1234',
    },
  },
};

export const createMockRider = (id, groupId, email) => {
  return {
    id,
    status: 'IDLE',
    group: groupId,
    user: {
      id: 2,
      email,
      user_type: 'rider',
      vehicle: {
        car_type: '',
        plate_no: '',
      },
    },
  };
};

export const createMockDriver = (id, groupId, email) => {
  return {
    id,
    status: 'IDLE',
    group: groupId,
    user: {
      id: 2,
      email,
      user_type: 'driver',
      vehicle: {
        car_type: 'benz',
        plate_no: '1234',
      },
    },
  };
};
