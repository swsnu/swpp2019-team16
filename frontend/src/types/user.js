import PropTypes from 'prop-types';

export const UserPropsTypes = PropTypes.shape({
  // rider/driver id
  id: PropTypes.number.isRequired,
  status: PropTypes.oneOf(['IDLE', 'ON_TAXI', 'DRIVING']).isRequired,
  // group id
  group: PropTypes.number,
  user: PropTypes.shape({
    // user id
    id: PropTypes.number.isRequired,
    email: PropTypes.string.isRequired,
    user_type: PropTypes.oneOf(['rider', 'driver']).isRequired,
    vehicle: PropTypes.shape({
      car_type: PropTypes.string.isRequired,
      plate_no: PropTypes.string.isRequired,
    }),
  }).isRequired,
});
