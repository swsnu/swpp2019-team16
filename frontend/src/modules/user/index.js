import user from './user';

export * from './user';
export default user;
