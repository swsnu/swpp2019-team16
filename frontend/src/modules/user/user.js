import { createAction, handleActions } from 'redux-actions';
import { takeLatest, put } from 'redux-saga/effects';
import * as userAPI from '../../lib/api/user';
import createRequestSaga, {
  createRequestActionTypes,
} from '../../lib/createRequestSaga';

const TEMP_SET_USER = 'user/TEMP_SET_USER';
const SOMEONE_ON_TAXI = 'user/SOMEONE_ON_TAXI';
const GO_TAXI = 'user/GO_TAXI';
const INIT_POINT_UPDATED_FLAG = 'user/INIT_POINT_UPDATED_FLAG';
const [CHECK, CHECK_SUCCESS, CHECK_FAILURE] = createRequestActionTypes(
  'user/CHECK',
);
const [LOGOUT, LOGOUT_SUCCESS, LOGOUT_FAILURE] = createRequestActionTypes(
  'user/LOGOUT',
);
const [
  UPDATE_POINT,
  UPDATE_POINT_SUCCESS,
  UPDATE_POINT_FAILURE,
] = createRequestActionTypes('user/UPDATE_POINT');
const [
  RIDER_ON_TAXI,
  RIDER_ON_TAXI_SUCCESS,
  RIDER_ON_TAXI_FAILURE,
] = createRequestActionTypes('user/RIDER_ON_TAXI');
const [
  DRIVER_GO_TAXI,
  DRIVER_GO_TAXI_SUCCESS,
  DRIVER_GO_TAXI_FAILURE,
] = createRequestActionTypes('user/DRIVER_GO_TAXI');

export const tempSetUser = createAction(TEMP_SET_USER, user => user);
export const check = createAction(CHECK, ({ id }) => ({ id }));
export const logout = createAction(LOGOUT);
export const updatePoint = createAction(UPDATE_POINT, ({ userId, point }) => ({
  userId,
  point,
}));

export const checkSaga = createRequestSaga(CHECK, userAPI.get);
export const riderOnTaxi = createAction(RIDER_ON_TAXI, ({ riderId }) => ({
  riderId,
}));
export const goTaxi = createAction(DRIVER_GO_TAXI, ({ driverId }) => ({
  driverId,
}));
export const someoneOnTaxi = createAction(SOMEONE_ON_TAXI, ({ riderId }) => ({
  riderId,
}));
export const taxiGone = createAction(GO_TAXI);
export const initPointUpdatedFlag = createAction(INIT_POINT_UPDATED_FLAG);

const riderOnTaxiSaga = createRequestSaga(RIDER_ON_TAXI, userAPI.riderOnTaxi);
const driverGoTaxiSaga = createRequestSaga(
  DRIVER_GO_TAXI,
  userAPI.driverGoTaxi,
);

function checkFailureSaga() {
  try {
    localStorage.removeItem('user');
  } catch (e) {
    console.error('localStorage is not working');
  }
}

export function* logoutSaga() {
  try {
    localStorage.removeItem('user');
    yield put({ type: LOGOUT_SUCCESS });
  } catch (e) {
    console.error(e);
    yield put({ type: LOGOUT_FAILURE });
  }
}

export const updatePointSaga = createRequestSaga(
  UPDATE_POINT,
  userAPI.updatePoint,
);
export function* userSaga() {
  yield takeLatest(CHECK, checkSaga);
  yield takeLatest(CHECK_FAILURE, checkFailureSaga);
  yield takeLatest(LOGOUT, logoutSaga);
  yield takeLatest(UPDATE_POINT, updatePointSaga);
  yield takeLatest(RIDER_ON_TAXI, riderOnTaxiSaga);
  yield takeLatest(DRIVER_GO_TAXI, driverGoTaxiSaga);
}

const initialState = {
  user: null,
  checkError: null,
  logoutError: null,
  updatePointError: null,
  onTaxiRidersList: [],
  driverGoTaxi: false,
  onTaxiError: null,
  goTaxiError: null,
  pointUpdated: false,
};

const user = handleActions(
  {
    [TEMP_SET_USER]: (state, { payload: user }) => ({
      ...state,
      user,
    }),
    [CHECK_SUCCESS]: (state, { payload: user }) => ({
      ...state,
      user,
      checkError: null,
    }),
    [CHECK_FAILURE]: (state, { payload: error }) => ({
      ...state,
      user: null,
      checkError: error,
    }),
    [LOGOUT_SUCCESS]: state => ({
      ...state,
      user: null,
      logoutError: null,
    }),
    [LOGOUT_FAILURE]: (state, { payload: error }) => ({
      ...state,
      logoutError: error,
    }),
    [UPDATE_POINT_SUCCESS]: (state, { payload: user }) => ({
      ...state,
      user: {
        ...user,
        user,
      },
      pointUpdated: true,
      updatePointError: null,
    }),
    [UPDATE_POINT_FAILURE]: (state, { payload: error }) => ({
      ...state,
      updatePointError: error,
    }),
    [RIDER_ON_TAXI_FAILURE]: (state, { payload: error }) => ({
      ...state,
      onTaxiError: error,
    }),
    [DRIVER_GO_TAXI_FAILURE]: (state, { payload: error }) => ({
      ...state,
      goTaxiError: error,
    }),
    [SOMEONE_ON_TAXI]: (state, { payload: riderId }) => ({
      ...state,
      onTaxiRidersList: [...state.onTaxiRidersList, riderId],
    }),
    [GO_TAXI]: state => ({
      ...state,
      driverGoTaxi: true,
    }),
    [INIT_POINT_UPDATED_FLAG]: state => ({
      ...state,
      pointUpdated: initialState.pointUpdated,
    }),
  },
  initialState,
);

export default user;
