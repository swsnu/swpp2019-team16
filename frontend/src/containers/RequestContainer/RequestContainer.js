import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Request from '../../components/Request';
import { withRouter } from 'react-router-dom';
import { requestCarpool } from '../../modules/carpoolRequest';

RequestContainer.propTypes = {};

function RequestContainer({ history }) {
  const dispatch = useDispatch();
  const user = useSelector(user => ({
    user: user.user,
  }));

  //TODO: it should be loaded from DB
  const fromList = ['SNU station', 'Nakseongdae station', 'Nokdu Street'];
  const toList = ['301 building', 'Student Center', 'Dormitory three-way'];
  const minimumPassenger = ['2', '3', '4'];

  const onClickRequest = useCallback(
    ({ userId, from, toString, minimumPassenger }) => {
      dispatch(requestCarpool({ userId, from, toString, minimumPassenger }));
      history.push('/waiting');
    },
    [dispatch, history],
  );

  if (!user) {
    return <div>we are loading user...</div>;
  }

  return (
    <Request
      user={user}
      fromList={fromList}
      toList={toList}
      minimumPassenger={minimumPassenger}
      onClickRequest={onClickRequest}
    />
  );
}

export default withRouter(RequestContainer);
