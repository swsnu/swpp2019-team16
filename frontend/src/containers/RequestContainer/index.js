import RequestContainer from './RequestContainer';
export default RequestContainer;
