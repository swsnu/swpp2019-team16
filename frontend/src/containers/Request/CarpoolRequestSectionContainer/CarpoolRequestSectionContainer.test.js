import React from 'react';
import { fireEvent, waitForElement } from '@testing-library/react';
import RequestContainer from './CarpoolRequestSectionContainer';
import { renderWithRedux } from '../../../test/utils';
import { MemoryRouter, Route } from 'react-router-dom';
import { mockRider } from '../../../types/__mock__/user';

jest.mock('../../../components/Request/CarpoolRequestSection', () =>
  jest.fn(props => (
    <div>
      <button
        onClick={() =>
          props.onCarpoolRequest({
            riderId: 1,
            from: 'A',
            to: 'B',
            minimumPassenger: '2',
          })
        }
      >
        request
      </button>
      <div>RequestPage</div>
    </div>
  )),
);

describe('<CarpoolRequestSectionContainer />', () => {
  const state = {
    user: {
      user: mockRider,
    },
  };

  it('SHOULD match with snapshot', async () => {
    const { container } = renderWithRedux(
      <MemoryRouter initialEntries={['/request']}>
        <Route component={RequestContainer} path="/request" />
      </MemoryRouter>,
      state,
    );

    expect(container).toMatchSnapshot();
  });

  it('should redirect to /login when user is not given', async () => {
    const _initialState = {
      ...state,
      user: {
        user: null,
      },
    };
    const MockLoginPage = () => <div>MockLoginPage</div>;
    const { getByText } = renderWithRedux(
      <MemoryRouter initialEntries={['/request']}>
        <Route component={RequestContainer} path="/request" />
        <Route component={MockLoginPage} path="/login" />
      </MemoryRouter>,
      _initialState,
    );

    await waitForElement(() => getByText('MockLoginPage'));

    expect(getByText('MockLoginPage')).toHaveTextContent('MockLoginPage');
  });

  it('should dispatch requestCarpool action when click request button', () => {
    const { getByText, store } = renderWithRedux(
      <MemoryRouter initialEntries={['/request']}>
        <Route component={RequestContainer} path="/request" />
      </MemoryRouter>,
      state,
    );

    fireEvent.click(getByText('request'));

    const fetchedActions = store.getActions();
    expect(fetchedActions.length).toBe(1);
    expect(fetchedActions[0]).toStrictEqual({
      type: 'carpoolRequest/REQUEST_CARPOOL',
      payload: {
        riderId: 1,
        from: 'A',
        to: 'B',
        minimumPassenger: '2',
      },
    });
  });

  it('should redirect to /waiting when click request button', async () => {
    const MockWaitingPage = () => <div>MockWaitingPage</div>;
    const { getByText } = renderWithRedux(
      <MemoryRouter initialEntries={['/request']}>
        <Route component={RequestContainer} path="/request" />
        <Route component={MockWaitingPage} path="/waiting" />
      </MemoryRouter>,
      state,
    );

    fireEvent.click(getByText('request'));

    await waitForElement(() => getByText('MockWaitingPage'));

    expect(getByText('MockWaitingPage')).toHaveTextContent('MockWaitingPage');
  });
});
