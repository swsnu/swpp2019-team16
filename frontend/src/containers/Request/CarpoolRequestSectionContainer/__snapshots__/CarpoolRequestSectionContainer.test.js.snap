// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarpoolRequestSectionContainer /> SHOULD match with snapshot 1`] = `
<div>
  <div>
    <button>
      request
    </button>
    <div>
      RequestPage
    </div>
  </div>
</div>
`;
